text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,268 @@
+
+Below is JavaScript example of how to think about performance instrumentation in the JavaScript SDKs (browser/node)
+
+```jsx
+// op is defined by https://develop.sentry.dev/sdk/performance/span-operations/
+// name has no specs but is expected to be low cardinality
+const transaction = Sentry.startTransaction({
+  op: "http.server",
+  name: "GET /",
+});
+
+// Need to set transaction on span so that integrations
+// can attach spans (I/O operations or framework-specific spans)
+// to the correct parent.
+Sentry.getCurrentHub().getScope().setSpan(transaction);
+
+// spans have a description, while transactions have names
+// op is an optional attribute, but a lot of the product relies on it
+// existing.
+// description technically should be low cardinality, but we have
+// no explicit documentation to say that it should (since spans
+// were not indexed at all for a while).
+const span = transaction.startChild({ description: "Long Task" });
+expensiveAction();
+span.finish();
+
+anotherAction();
+
+const secondSpan = transaction.startChild({ description: "Another Task" });
+
+// transaction + all child spans sent to Sentry only when `finish()` is called.
+transaction.finish();
+Sentry.getCurrentHub().getScope().setSpan(undefined);
+
+// second span info is not sent to Sentry because transaction is already finished.
+secondSpan.finish();
+```
+
+In our integrations that add automatic instrumentation, things look something like so:
+
+```jsx
+const parentSpan = Sentry.getCurrentHub().getSpan();
+
+// parentSpan can be undefined if no span is on the scope, this leads to
+// child span just being lost
+const span = parentSpan?.startChild({ description: "something" });
+
+Sentry.getCurrentHub().getScope().setSpan(span);
+
+work();
+
+span.finish();
+
+// span is finished, so parent is put back onto scope
+Sentry.getCurrentHub().getScope().setSpan(parentSpan);
+```
+
+Most users do the above also when nested within their application, as often you assume that a transaction is defined that you can attach (very common in a web server context).
+
+To add instrumentation to their applications, users have to know concepts about hubs/scopes/transactions/spans and understand all the different nuances and use cases. This can be difficult, and presents a big barrier to entry for new users.
+
+Also, since transactions/spans are different classes (span is a parent class of transaction), users have to understand the impacts that the same method will have on both the transaction/span. For example, currently calling `setTag` on a transaction will add a tag to the transaction event (which is searchable in Sentry), while `setTag` on a span just adds it to the span, and the field is not searchable. `setData` on a span adds it to `span.data`, while `setData` on a transaction is undefined behaviour (some SDKs throw away the data, others add it to `event.extras`).
```

Comment:
Isn't it the other way around? the transaction is a parent class of span, at least on Mobile.